explorer/src/indexer.rs

```diff
@@ -0,0 +1,65 @@
+use crate::db::ExplorerDb;
+use chain_impl_mockchain::block::Block;
+use chain_impl_mockchain::block::HeaderId as HeaderHash;
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::sync::Mutex;
+
+#[derive(Debug, Error)]
+pub enum IndexerError {
+    #[error("could not deserialize response")]
+    CannotDeserialize(#[from] serde_json::Error),
+    #[error("io error")]
+    IOError(#[from] std::io::Error),
+    #[error("hash error")]
+    HashError(#[from] chain_crypto::hash::Error),
+    #[error("url error")]
+    UrlError(#[from] url::ParseError),
+    #[error(transparent)]
+    DbError(#[from] crate::db::error::ExplorerError),
+}
+
+#[derive(Clone)]
+pub struct Indexer {
+    pub db: ExplorerDb,
+    tip_candidate: Arc<Mutex<Option<HeaderHash>>>,
+}
+
+impl Indexer {
+    pub fn new(db: crate::db::ExplorerDb) -> Self {
+        let tip_candidate = Arc::new(Mutex::new(None));
+        Indexer { db, tip_candidate }
+    }
+
+    pub async fn apply_block(&self, block: Block) -> Result<(), IndexerError> {
+        tracing::info!("applying {}", block.header().id());
+
+        // TODO: technically this could dispatch a task, as there is a possibility of applying
+        // blocks (siblings) in parallel, but that is a mission for another day.  biggest concern
+        // is that the we receive two consecutive blocks, if the first is really big and costly to
+        // apply, we may try to apply the next one too soon...
+        let _state_ref = self.db.apply_block(block.clone()).await?;
+
+        let mut guard = self.tip_candidate.lock().await;
+        if guard
+            .map(|hash| hash == block.header().id())
+            .unwrap_or(false)
+        {
+            let hash = guard.take().unwrap();
+            self.set_tip(hash).await;
+        }
+
+        Ok(())
+    }
+
+    pub async fn set_tip(&self, tip: HeaderHash) {
+        let successful = self.db.set_tip(tip).await;
+
+        if !successful {
+            let mut guard = self.tip_candidate.lock().await;
```

Comment:
Well, it's kinda hacky I guess. 

We don't have any guarantees about the order of events, and if we set a block as the tip, but haven't indexed the block yet, queries in between will fail.

In the past, inside the node, we broadcasted tip updates before broadcasting the blocks. I'm not sure if that changed actually, because I believe you refactored that, but I wanted the explorer to keep working regardless of the order.

And here, because the tip and block subscriptions are different streams, I think it may be possible for this to happen regardless of the order they are emitted in jormungandr.

Reply:
Well, I changed the return type to a `Result<(), BlockNotFound>` instead, as it is a bit more self-documenting.